netket/graph/continuous_geometry.py

```diff
@@ -0,0 +1,146 @@
+    """Documentation"""
+
+    def __init__(
+        self,
+        pbc: Union[bool, tuple[bool, ...]],
+        dim: Optional[int] = None,
+        lattice: Optional = None,
+    ):
+        """
+        Constructs new ``Particles`` given specifications
+        of the continuous space they are defined in.
+
+        Args:
+            pbc: Tuple or bool indicating whether to use periodic boundary conditions in a given physical dimension.
+                If tuple it must have the same length as domain. If bool the same value is used for all the dimensions
+                defined in domain.
+            dim: (Optional) The number of spatial dimensions in the Hilbert space
+            lattice: (Optional) The lattice that describes the Hilbert space. For 'pbc=True' this is the basis of the
+                        physical simulation cell. For 'pbc=False' the standard Euclidean basis is assumed.
+            geometry: (Optional) A geometry object. Either a periodic box 'Cell' or free space 'Free'.
+        """
+        if dim is None and lattice is None:
+            if isinstance(pbc, bool):
+                raise ValueError(
+                    """If 'dim' and 'lattice' is not given, the 'pbc' has to be a tuple to infer the spatial
+                    dimensions"""
+                )
+            dim = len(pbc)
+
+        elif lattice is None:
+            self._dim = dim
+        else:
+            self._dim = jnp.shape(lattice)[0]
+
+        if isinstance(pbc, bool):
+            self._pbc = (pbc,) * self._dim
+
+        if lattice is None:
+            # assume cubic box if lattice is not given
+            self._lat = jnp.eye(self._dim)
+        else:
+            self._lat = jnp.array(lattice)
+
+        self._ilat = jnp.linalg.inv(self._lat)
+
+        if not self._dim == len(self._pbc) == jnp.shape(self._lat)[0]:
+            raise ValueError(
+                """There is a mismatch between the number of spatial dimensions
+                inferred by 'pbc', 'lattice' and 'dim."""
+            )
+        self._hash = None
+
+    @property
+    def pbc(self) -> Tuple:
+        r"""Whether or not periodic boundary conditions are applied along a spatial dimension (x,y,z,...)"""
+        return self._pbc
+
+    @property
+    def dim(self):
+        r"""The spatial dimensions of the Hilbert space"""
+        return self._dim
+
+    @property
+    def lattice(self):
+        r"""The lattice used."""
+        return self._lat
+
+    @property
+    def inv_lattice(self):
+        r"""The reciprocal lattice to the lattice."""
+        return self._ilat
+
+    @property
+    def volume(self) -> int:
+        r"""The volume of the Hilbert space if defined (only defined if it is fully periodic)"""
+        raise NotImplementedError
+
+    @property
+    def extent(self):
+        r"""Spatial extension in each spatial dimension"""
+        raise NotImplementedError
+
+    def distance(self, x, y=None, norm=False, mode=None) -> list[list]:
```

Comment:
indeed what is norm? `mode` should be enough I guess (e.g. `mode=L2` )

Reply:
If y=None it means distance(x,x) with the caveat that the norm is computed differently in that case.

To not having to call the method twice in case you want the norm and the vectors, there is the extra flag `norm`. We could have a seperate mode keyword for that.

Reply:
not if x contains N different positions. x can have shape (nsamples, nparticles, sdim) then only the diagonal is 0.

Reply:
So in the case  of x being a batch of samples, you return a matrix `nsamples x nsamples` of distances?

Can't we do this by having the following signature for the distance function? 
 - `distance( x:Array[N], y: Array[N]) -> Array[] # (a scalar)`   
 - `distance( x:Array[B1, N], y: Array[N]) -> Array[B1]` 
 - `distance( x:Array[N], y: Array[B2,N]) -> Array[B2]` 
 - `distance( x:Array[B1, N], y: Array[B2,N]) -> Array[B1, B2]` 
 - `distance( x:Array[B1, B2, N], y: Array[B3, N]) -> error`